languages/python/reference/exercise-concepts/robot-simulator.md

```diff
@@ -0,0 +1,62 @@
+
+- `property()` built-in:  a function that returns a property attribute.  When used as a decorator, this transforms the passed-in method into a _getter_ method for read-only attribute with the same name and docstring.
+
+- assignment:  the example uses assignment for all the instance properties and `instructions` dictionary.
+
+- instance properties: This exercise rquires one or more instance properties to persist passed in data.
+
+- mutability: In the example, knowing there are no protected or private properties in python and so conciously mutating `self.x`, `self.y` and `self.compass` through the called instance methods.
+
+- method arguments: the example `__init__` method needs to take `self`, direction, x, and y (coordinates) as arguments.  It also uses `self` and `commands` (a string) for arguments to the `move()` method.
+- default arguments:  Pre-setting a function argument value to protect against it not being passed..  The example uses `direction = NORTH` and `x=0, y=0` to ensure those values for a robot even if they are not passed in initally.
+
+- dictionaries/mapping type: the example uses a dictionary to map paassed in move arguemtns to methods that perform the moves.  The example also uses a dictionary/mapping created by calling `str.maketrans()`.
+
+- bracket notation in context of dictionaries: the example uses passed in move arguments as keys via bracket notation to look up corresponding methods for moving the robot in the _insructions_ dictionary. 
+
+- iteration/iterators/iterating: the example uses a `for loop` to iterate through the passed in command string.
+
+- composition:  adding functionality from a class by incorporating and instance of that class in a class you are creating.  The example cureates a `robot` by instatiating a `compass`, and assiging it to the `self`.compass property of `robot`.
+
+- modulous operator: the example uses the modulus operator to calculate the appropriate _compass_ setting when calling the `right` compass method.
+
+- lists:  the example stores compass direction constants in a `list`
+
+- for loop: the example uses a `for loop` to iterate through the letters of the passed-in `commands` string and looks up the corresponding values in a dictionary, so that the appropriate methods can be called to move the `robot`.
+
+- callable objects:  knowing that appending `()` to the name of an instance method _calls_ it, since instances of classes are _callable_, and implement `__call__()`.
+
+- compairison operator built-in (`==` or equal) :  `==` calls the magic method `__equ__()`.  Knowing that objects of different types never compare equal unless they are numerica types.  `==` is always defined, but for some object types (_like class objects_) it is equivelent to calling `is()`.  This excercise uses the compairison operator to test that the `self.direction` attribute is equal to one of the constant values defined.
```

Comment:
Spelling: "compairison" -> "comparison", "numerica" -> "numerical", "equivelent" -> "equivalent", "compairison" -> "comparison". And that should be `__eq__` rather than `__equ__`.  Also `==` is the _equality_ operator (to distinguish it from the other comparison operators), and it is not truly a builtin (it doesn't appear in the builtins module), rather it's defined in the syntax of Python itself, more like a keyword.

Also the line "[k]nowing that objects of different [...] numerical types" is a bit confusing ... maybe that should be "By default objects of different types are never considered equal to each other unless they are numerical types (ie int, float) or have otherwise overloaded the default implementation of the `__eq__` magic method." ... 

The algorithm when called on two terms of different types (including numerical types) is a bit more complicated than described -- for `a == b` it will call `a.__eq__(b)` and if that returns `NotImplemented` it will call `b.__eq__(a)` and if _that_ returns NotImplemented, then it will finally fall back on the default implementation on `object` itself, which will return `a is b` -- but I'm not sure we necessarily want to go into that, although perhaps a link to the [standard docs](https://docs.python.org/3/reference/datamodel.html#object.__eq__)?

IIRC correctly the default `object.__eq__(a, b)` is actually implemented as `return a is b`, so unless `__eq__` is defined on a or b (or one of their superclasses) then `a == b` is exactly the same as `a is b` in all cases, not just class objects. Also it's a bit odd to say "calling `is()`" since `is` isn't a function and cannot be called.

Reply:
Agreed that a link to the docs would be good.  I think my mistake was using the [Standard Types](https://docs.python.org/3/library/stdtypes.html#comparisons) docs for explanation instead of looking at the docs for `__eq__`.    Revised wording:

> - equality operator (`==`) :  `==` calls the magic method `__eq__()`.  By default, objects of different types are never considered equal to each other unless they are numerical types (ie int, float) or have otherwise overloaded the default implementation of the `__eq__` magic method.  `==` is always defined, but for some object types (_like class objects_) it is equivalent to calling `is`.  See [`__eq__`](https://docs.python.org/3/reference/datamodel.html#object.__eq__)  for additional details.  This exercise uses the equality operator to test that the `self.direction` attribute is equal to one of the constant values defined.

Reply:
Much better.